pkg/util/gate/custom_resource_gate.go

```diff
@@ -0,0 +1,78 @@
+package gate
+
+import (
+	"os"
+	"strings"
+
+	"github.com/openkruise/kruise/pkg/apis"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+const (
+	envCustomResourceEnable = "CUSTOM_RESOURCE_ENABLE"
+)
+
+var (
+	internalScheme  = runtime.NewScheme()
+	discoveryClient discovery.DiscoveryInterface
+)
+
+func init() {
+	_ = apis.AddToScheme(internalScheme)
+	cfg, err := config.GetConfig()
+	if err != nil {
+		klog.Errorf("unable to set up client config: %v", err)
+		os.Exit(1)
+	}
+	discoveryClient = discovery.NewDiscoveryClientForConfigOrDie(cfg)
+}
+
+// ResourceEnabled help runnable check if the custom resource is valid and enabled
+// 1. If this CRD is not found from kueb-apiserver, it is invalid.
+// 2. If 'CUSTOM_RESOURCE_ENABLE' env is not empty and this CRD kind is not in
```

Comment:
2.   .... this CRD kind is not in $CUSTOM_RESOURCE_ENABLE.

Reply:
done